### PR TITLE
Jetty 9.4.x 3550 queued thread pool stalled

### LINKED ISSUE #3550: Server becomes unresponsive after sitting idle from a load spike
I tested Jetty 9.4.16 over the weekend with quite some load, although if's currently only available on Maven central and not officially been announced.

After a few hours, Jetty became unresponsive and I got 502 errors from my reverse proxy before it. I downgraded to Jetty 9.4.15 again and it seems to be running fine.

I really can't pinpoint what might cause this behavior, as it only happened after a few hours - but repeatedly. I use Glowroot to monitor the application and I found nothing obvious in the threaddump either. Out of ideas how to debug such situations.

Can somebody reproduce this? Any ideas how to further debug this?

### PR DESCRIPTION
**Pull request by gregw**
This fixes #3550 by ensure that a new thread is always started when:
 + a task is executed, the queue is not empty and there are no idle threads available (this may race, but will only start addition threads)
 + if a thread exits and the queue is not empty

### CODE REVIEW

File: jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java [hunk 1]
```diff
@@ -368,10 +498,99 @@ public void testConstructorMinMaxThreadsValidation()
         });
     }
 
+    @Test
+    public void testDump() throws Exception
+    {
+        QueuedThreadPool pool = new QueuedThreadPool(4, 3);
+
+        String dump = pool.dump();
+        // TODO use hamcrest 2.0 regex matcher
```
Comment (joakime) [inline jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java]: I need to do a refresh of maven-test-helper for Jetty 9.4.x anyway.
In the meantime you can use the maven-test-helper regex version ...

``` java
import static org.eclipse.jetty.websocket.common.test.MoreMatchers.regex;

assertThat(endpoint.capture.pollMessages(), regex("^onBinary\\(.*InputStream.*"));
```

Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java]: never mind, that's a websocket specific matcher.
i'll upgrade maven-test-helper today. (and even include the new `Net` helper for junit 5. 879e161c22a3613986e7ffd1601fd2711bba556a)

Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java]: thanks

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java [hunk 2]
```diff
@@ -603,7 +604,7 @@ public void dump(Appendable out, String indent) throws IOException
             String knownMethod = "";
             for (StackTraceElement t : trace)
             {
-                if ("idleJobPoll".equals(t.getMethodName()) && t.getClassName().endsWith("QueuedThreadPool"))
+                if ("idleJobPoll".equals(t.getMethodName()) && t.getClassName().endsWith("QueuedThreadPool$Runner"))
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java]: can this be ...

``` java
if ("idleJobPoll".equals(t.getMethodName()) && t.getClassName().equals(Runner.class.getName()))
```

instead?

**Pull request merged**